### PR TITLE
[SPARK-44788][CONNECT][PYTHON][SQL] Add from_xml and schema_of_xml to pyspark, spark connect and sql function

### PR DESCRIPTION
**Pull request by sandip-db**
### What changes were proposed in this pull request?
Add from_xml and schema_of_xml to pyspark, spark connect and sql function

### Why are the changes needed?
from_xml parses XML data nested in a `Column` into a struct. schema_of_xml infers schema from XML data in a `Column`.
This PR adds these two functions to pyspark, spark connect and SQL function registry.
It is one of the series of PR to add native support for [XML File Format](https://issues.apache.org/jira/browse/SPARK-44265) in spark.

### Does this PR introduce _any_ user-facing change?
Yes, it adds from_xml and schema_of_xml to pyspark, spark connect and sql function

### How was this patch tested?
- Added new unit tests
- Github Action

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -7367,15 +7367,83 @@ object functions {
...
+   *                Xml data source.
+   *                See
+   *                <a href=
+   *                "https://spark.apache.org/docs/latest/sql-data-sources-xml.html#data-source-option">
+   *                Data Source Option</a> in the version you use.
+   * @group collection_funcs
+   * @since 4.0.0
+   */
+  // scalastyle:on line.size.limit
+  def from_xml(e: Column, schema: String, options: Map[String, String]): Column = {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:7418]: Let's probably remove the Scala specific `Map`, and only have one with Java signature. Scala can easily use them.

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -13041,6 +13041,120 @@ def json_object_keys(col: "ColumnOrName") -> Column:
...
+        for the version you use.
+
+        .. # noqa
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        a new column of complex type from given XML object.
+
+    Examples
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:13075]: I think we should improve the examples here but let's do that separately. cc @allisonwang-db and @zhengruifeng FYI

File: python/pyspark/sql/tests/connect/test_connect_function.py [hunk 3]
```diff
@@ -1821,6 +1821,106 @@ def test_json_functions(self):
...
+        # |<p><a>1</a></p>|<p><a>4</a><a>5</a><a>6</a></p>|          <p><a attr="t"></a></p>|
+        # +---------------+-------------------------------+---------------------------------+
+
+        cdf = self.connect.sql(query)
+        sdf = self.spark.sql(query)
+
+        # test from_xml
+        for schema in [
+            "a INT",
+#            StructType([StructField("a", IntegerType())]),
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/connect/test_connect_function.py:1844]: Let's probably remove commented codes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 4]
```diff
@@ -830,7 +830,11 @@ object FunctionRegistry {
     // csv
     expression[CsvToStructs]("from_csv"),
     expression[SchemaOfCsv]("schema_of_csv"),
-    expression[StructsToCsv]("to_csv")
+    expression[StructsToCsv]("to_csv"),
+
+    // Xml
+    expression[XmlToStructs] ("from_xml"),
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:836]:
```suggestion
    expression[XmlToStructs]("from_xml"),
```

Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:836]: To register this properly, I think `XmlToStructs` has to be decorated by `ExpressionDescription` like you did in `SchemaOfXml`.

File: python/pyspark/sql/functions.py [hunk 5]
```diff
@@ -13041,6 +13041,120 @@ def json_object_keys(col: "ColumnOrName") -> Column:
...
+
+    Parameters
+    ----------
+    col : :class:`~pyspark.sql.Column` or str
+        a column or column name in XML format
+    schema : :class:`StructType` or str
+        a StructType or Python string literal with a DDL-formatted string
+        to use when parsing the Xml column
+    options : dict, optional
+        options to control parsing. accepts the same options as the Xml datasource.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:13064]:
```suggestion
        options to control parsing. Accepts the same options as the Xml datasource.
```

File: python/pyspark/sql/functions.py [hunk 6]
```diff
@@ -13041,6 +13041,120 @@ def json_object_keys(col: "ColumnOrName") -> Column:
...
+    xml : :class:`~pyspark.sql.Column` or str
+        a XML string or a foldable string column containing a XML string.
+    options : dict, optional
+        options to control parsing. accepts the same options as the XML datasource.
+        See `Data Source Option <https://spark.apache.org/docs/latest/sql-data-sources-xml.html#data-source-option>`_
+        for the version you use.
+
+        .. # noqa
+
+        .. versionchanged:: 4.0.0
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:13126]: You can remove this since this is a new feature, we won't need to annotate them.

**Comment by HyukjinKwon** [comment]
cc @itholic mind helping review this please?

File: python/pyspark/sql/tests/connect/test_connect_function.py (hunk 3 above)
Comment (sandip-db) [reply python/pyspark/sql/tests/connect/test_connect_function.py:1844]: I would like to support this. pyspark sql works fine, but pyspark connect is failing to parse the StructType.
Any pointers would help.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/FunctionTestSuite.scala [hunk 7]
```diff
@@ -229,6 +229,18 @@ class FunctionTestSuite extends ConnectFunSuite {
     schema_of_csv("x,y"),
     schema_of_csv(lit("x,y"), Collections.emptyMap()))
   testEquals("to_csv", to_csv(a), to_csv(a, Collections.emptyMap[String, String]))
+  testEquals(
+    "from_xml",
+    from_xml(a, schema),
+    from_xml(a, lit(schema.toDDL)),
+    // from_xml(a, lit(schema.json)),
```
Comment (sandip-db) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/FunctionTestSuite.scala:236]: @HyukjinKwon @itholic 
This is failing due to a parse error. So I commented it temporarily. The same passes for from_json() above in this test.
Can you please share some pointers?

File: python/pyspark/sql/functions.py (hunk 2 above)
Comment (allisonwang-db) [reply python/pyspark/sql/functions.py:13075]: Documentation is extremely important for a better user-experience. @sandip-db could you please create a ticket under https://issues.apache.org/jira/browse/SPARK-44728.

File: python/pyspark/sql/functions.py [hunk 8]
```diff
@@ -13041,6 +13041,117 @@ def json_object_keys(col: "ColumnOrName") -> Column:
...
+    Parses a column containing a XML string to a row with
+    the specified schema. Returns `null`, in the case of an unparseable string.
+
+    .. versionadded:: 4.0.0
+
+    Parameters
+    ----------
+    col : :class:`~pyspark.sql.Column` or str
+        a column or column name in XML format
+    schema : :class:`StructType` or str
```
Comment (itholic) [inline python/pyspark/sql/functions.py:13060]:
```diff
- schema : :class:`StructType` or str
+ schema : :class:`StructType`, :class:`~pyspark.sql.Column` or str
```

File: python/pyspark/sql/functions.py (hunk 2 above)
Comment (itholic) [reply python/pyspark/sql/functions.py:13075]: Yeah, we can do it in separate PR, but let's simply put some newline to readability for now.

e.g.

```python
    >>> from pyspark.sql.types import *
    >>> from pyspark.sql.functions import from_xml, schema_of_xml, lit

    StructType input with simple IntegerType.

    >>> data = [(1, '''<p><a>1</a></p>''')]
    >>> df = spark.createDataFrame(data, ("key", "value"))
    >>> schema = StructType([StructField("a", IntegerType())])
    >>> df.select(from_xml(df.value, schema).alias("xml")).collect()
    [Row(xml=Row(a=1))]

    String input.

    >>> df.select(from_xml(df.value, "a INT").alias("xml")).collect()
    [Row(xml=Row(a=1))]

    StructType input with complex ArrayType.

    >>> data = [(1, '<p><a>1</a><a>2</a></p>')]
    >>> df = spark.createDataFrame(data, ("key", "value"))
    >>> schema = StructType([StructField("a", ArrayType(IntegerType()))])
    >>> df.select(from_xml(df.value, schema).alias("xml")).collect()
    [Row(xml=Row(a=[1, 2]))]

    Column input generated by schema_of_xml.

    >>> schema = schema_of_xml(lit(data[0][1]))
    >>> df.select(from_xml(df.value, schema).alias("xml")).collect()
    [Row(xml=Row(a=[1, 2]))]
```

File: python/pyspark/sql/tests/connect/test_connect_function.py (hunk 3 above)
Comment (itholic) [reply python/pyspark/sql/tests/connect/test_connect_function.py:1844]: Then let's create JIRA and comment it as TODO for now. I think we need separate investigation.

e.g.
```python
        # TODO(SPARK-XXXXX): description (Please fill the actual description and JIRA number)
        for schema in [
            "a INT",
            # StructType([StructField("a", IntegerType())]),
            # StructType([StructField("a", ArrayType(IntegerType()))]),
        ]:
```

File: python/pyspark/sql/tests/connect/test_connect_function.py [hunk 9]
```diff
@@ -1821,6 +1821,106 @@ def test_json_functions(self):
...
+                sdf.select(SF.from_xml(sdf.a, schema, {"mode": "FAILFAST"})),
+            )
+            self.compare_by_show(
+                cdf.select(CF.from_xml("a", schema, {"mode": "FAILFAST"})),
+                sdf.select(SF.from_xml("a", schema, {"mode": "FAILFAST"})),
+            )
+
+        for schema in [
+            "STRUCT<a: ARRAY<INT>>",
+#            StructType([StructField("a", ArrayType(IntegerType()))]),
```
Comment (itholic) [inline python/pyspark/sql/tests/connect/test_connect_function.py:1866]: ditto. And let's move the `#` for the proper place.
```diff
- #            StructType([StructField("a", ArrayType(IntegerType()))]),
+              # StructType([StructField("a", ArrayType(IntegerType()))]),
```

File: python/pyspark/sql/tests/test_functions.py [hunk 10]
```diff
@@ -1286,6 +1281,27 @@ def test_from_csv(self):
             message_parameters={"arg_name": "schema", "arg_type": "int"},
         )
 
+    def test_schema_of_xml(self):
```
Comment (itholic) [inline python/pyspark/sql/tests/test_functions.py:1284]: I think we might need to test for non-connect as well??

Seems like we only compare between connect <> non-connect from `python/pyspark/sql/tests/connect/test_connect_function.py` by using `compare_by_show `, but there is no test for data correctness.

**Review by itholic — COMMENTED** [review]
Left some comments for Python side, otherwise the overall structure of the implementation itself seems fine to me.

File: python/pyspark/sql/functions.py (hunk 2 above)
Comment (sandip-db) [reply python/pyspark/sql/functions.py:13075]: > could you please create a ticket
@allisonwang-db 
Sure, created a ticket here: https://issues.apache.org/jira/browse/SPARK-45186

**Review by HyukjinKwon — APPROVED** [review]
cc @zhengruifeng too

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/FunctionTestSuite.scala (hunk 7 above)
Comment (HyukjinKwon) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/FunctionTestSuite.scala]: Using JSON as its schema is discouraged actually. I think you can just remove it for now.

Comment (sandip-db) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/FunctionTestSuite.scala]: It works now :-)

File: python/pyspark/sql/tests/connect/test_connect_function.py [hunk 11]
```diff
@@ -1821,6 +1821,111 @@ def test_json_functions(self):
...
+        # +---------------+-------------------------------+---------------------------------+
+
+        cdf = self.connect.sql(query)
+        sdf = self.spark.sql(query)
+
+        # test from_xml
+        for schema in [
+            "a INT",
+            # StructType([StructField("a", IntegerType())]),
+            # StructType([StructField("a", ArrayType(IntegerType()))]),
```
Comment (zhengruifeng) [inline python/pyspark/sql/tests/connect/test_connect_function.py:1847]: are they supported?

File: python/pyspark/sql/tests/connect/test_connect_function.py [hunk 12]
```diff
@@ -1821,6 +1821,111 @@ def test_json_functions(self):
...
+                sdf.select(SF.from_xml(sdf.a, schema, {"mode": "FAILFAST"})),
+            )
+            self.compare_by_show(
+                cdf.select(CF.from_xml("a", schema, {"mode": "FAILFAST"})),
+                sdf.select(SF.from_xml("a", schema, {"mode": "FAILFAST"})),
+            )
+
+        for schema in [
+            "STRUCT<a: ARRAY<INT>>",
+            # StructType([StructField("a", ArrayType(IntegerType()))]),
```
Comment (zhengruifeng) [inline python/pyspark/sql/tests/connect/test_connect_function.py:1868]: is this supported?

File: python/pyspark/sql/tests/connect/test_connect_function.py (hunk 11 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/tests/connect/test_connect_function.py:1847]: Had a quick offline sync. He will file a JIRA and take a separate look.

Comment (sandip-db) [reply python/pyspark/sql/tests/connect/test_connect_function.py:1847]: Here is the jira: 
https://issues.apache.org/jira/browse/SPARK-45190

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**